src/clj/puppetlabs/services/jruby/jruby_internal.clj

```diff
@@ -139,39 +146,42 @@
 (schema/defn ^:always-validate
   cleanup-pool-instance!
   "Cleans up and cleanly terminates a JRubyInstance and removes it from the pool."
-  [{:keys [scripting-container pool] :as instance} :- JRubyInstance]
+  [{:keys [scripting-container pool] :as instance} :- jruby-schemas/JRubyInstanceSchema
+   shutdown-fn :- IFn]
   (.unregister pool instance)
-  ;; TODO: need to add support for a callback hook, so that consumers like
-  ;; puppet-server can do their own cleanup.
-  ;(.terminate jruby-puppet)
+  (shutdown-fn instance)
   (.terminate scripting-container)
   (log/infof "Cleaned up old JRubyInstance with id %s." (:id instance)))
 
 (schema/defn ^:always-validate
-  create-pool-instance! :- JRubyInstance
+  create-pool-instance! :- jruby-schemas/JRubyInstanceSchema
   "Creates a new JRubyInstance and adds it to the pool."
   [pool :- jruby-schemas/pool-queue-type
    id :- schema/Int
    config :- jruby-schemas/JRubyConfig
    flush-instance-fn :- IFn]
-  (let [{:keys [ruby-load-path gem-home compile-mode]} config]
+  (let [{:keys [ruby-load-path gem-home compile-mode lifecycle]} config
+        init-fn (:initialize lifecycle)
+        init-env-vars-fn (:initialize-env-variables lifecycle)]
```

Comment:
Didn't @camlow325 come up with some idea about how we could handle the environment vars that wouldn't require us to make it an official lifecycle fn?  Like, I think maybe he was suggesting that we just always do our default environment var initialization, and then document how a user can override it as part of their implementation of their `init` fn?  My memory is rusty but I feel like he'd come up with some fairly compelling argument for doing it that way.